File: test/jv-stub/src/test/java/runtime/AnyEnvelopeCodecTest.java
package runtime;

import baboon.runtime.shared.*;
import baboon.runtime.shared.BaboonAnyOpaque.*;
import com.fasterxml.jackson.databind.node.NullNode;
import org.junit.jupiter.api.Test;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import static org.junit.jupiter.api.Assertions.*;

public class AnyEnvelopeCodecTest {
    @Test
    public void nativeFormatsRoundTripAllKindsAndLeaveTheTrailer() throws Exception {
        for (byte kind : new byte[] {0, 1, 2, 3, 6, 7}) {
            var meta = new AnyMeta(kind, (kind & 4) != 0 ? "domain" : null, (kind & 2) != 0 ? "1.0.0" : null, (kind & 1) != 0 ? "type" : null);
            var json = new AnyOpaqueJson(meta, NullNode.getInstance());
            assertEquals(json, BaboonAnyJsonCodec.decode(kind, BaboonAnyJsonCodec.encode(BaboonCodecContext.Compact, kind, null, null, null, json)));
            var binary = new AnyOpaqueUeba(meta, new byte[] {(byte) 0xaa, (byte) 0xbb});
            var buffer = new ByteArrayOutputStream();
            var writer = new LEDataOutputStream(buffer);
            BaboonAnyBinCodec.encode(BaboonCodecContext.Compact, writer, kind, null, null, null, binary);
            writer.writeByte(42);
            if (kind == 0) assertArrayEquals(new byte[] {7,0,0,0,1,0,0,0,0,(byte) 0xaa,(byte) 0xbb,42}, buffer.toByteArray());
            var reader = new LEDataInputStream(new ByteArrayInputStream(buffer.toByteArray()));
            assertEquals(binary, BaboonAnyBinCodec.decode(reader, kind));
            assertEquals(42, reader.readByte());
        }
    }

    @Test
    public void consumesMetadataExtensionsAndRejectsTruncatedPayloads() throws Exception {
        var extended = new LEDataInputStream(new ByteArrayInputStream(new byte[] {8,0,0,0,3,0,0,0,0,17,34,51,42}));
        assertArrayEquals(new byte[] {51}, BaboonAnyBinCodec.decode(extended, (byte) 0).bytes());
        assertEquals(42, extended.readByte());
        var truncated = new LEDataInputStream(new ByteArrayInputStream(new byte[] {8,0,0,0,3,0,0,0,0,17}));
        var failure = assertThrows(BaboonCodecException.DecoderFailure.class, () -> BaboonAnyBinCodec.decode(truncated, (byte) 0));
        assertTrue(failure.getMessage().contains("short read while skipping meta-extension bytes"));
    }
}

File: test/kt-stub-kmp/src/test/kotlin/runtime/AnyEnvelopeCodecTest.kt
package runtime

import baboon.runtime.shared.*
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class AnyEnvelopeCodecTest {
    private val meta = AnyMeta(0, null, null, null)
    private val payload = byteArrayOf(1, 2)

    private fun encode(value: AnyOpaque, kind: Byte): ByteArray {
        val writer = BaboonBinaryWriter()
        BaboonAnyBinCodec.encode(BaboonCodecContext.Compact, writer, kind, null, null, null, value)
        return writer.toByteArray()
    }

    private fun decode(bytes: ByteArray, kind: Byte): AnyOpaqueUeba =
        BaboonAnyBinCodec.decode(BaboonBinaryReader(bytes), kind)

    @Test
    fun binaryEnvelopeRetainsExactBytes() {
        val expected = byteArrayOf(7, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2)
        assertContentEquals(expected, encode(AnyOpaqueUeba(meta, payload), 0))
        assertEquals(AnyOpaqueUeba(meta, payload), decode(expected, 0))
    }

    @Test
    fun binaryEnvelopeSkipsFutureMetadataExtensions() {
        val extended = byteArrayOf(9, 0, 0, 0, 3, 0, 0, 0, 0, 42, 43, 1, 2)
        assertEquals(AnyOpaqueUeba(meta, payload), decode(extended, 0))
    }

    @Test
    fun jsonEnvelopeRetainsContentAndChecksKind() {
        val content = JsonObject(mapOf("x" to JsonPrimitive(1)))
        val value = AnyOpaqueJson(meta, content)
        val expected = JsonObject(mapOf("${'$'}ak" to JsonPrimitive(0), "${'$'}c" to content))
        assertEquals(expected, BaboonAnyJsonCodec.encode(BaboonCodecContext.Compact, 0, null, null, null, value))
        assertEquals(value, BaboonAnyJsonCodec.decode(0, expected))
        assertFailsWith<BaboonCodecException.DecoderFailure> { BaboonAnyJsonCodec.decode(1, expected) }
        assertFailsWith<BaboonCodecException.DecoderFailure> {
            BaboonAnyJsonCodec.decode(0, JsonObject(mapOf("${'$'}ak" to JsonPrimitive(0))))
        }
    }

    @Test
    fun encodingRetainsKindAndFacadeRequirements() {
        val binary = AnyOpaqueUeba(meta, payload)
        assertFailsWith<BaboonCodecException.EncoderFailure> { encode(binary, 1) }
        assertFailsWith<BaboonCodecException.EncoderFailure> {
            BaboonAnyJsonCodec.encode(BaboonCodecContext.Compact, 0, null, null, null, binary)
        }
        assertFailsWith<BaboonCodecException.EncoderFailure> {
            encode(AnyOpaqueJson(meta, JsonPrimitive(1)), 0)
        }
    }
}
